src/Game/Game.elm

```diff
@@ -134,6 +140,56 @@ update msg model =
             ( { model | windowSize = size }, Cmd.none )
 
 
+updateViewportOffset : Vector -> Model -> Vector
+updateViewportOffset prevPosition ({ windowSize, viewportX, viewportY, maps } as model) =
+    let
+        tileSize = 32
+
+        ( prevX, prevY ) =
+            Vector.scale tileSize prevPosition
+
+        ( x, y ) =
+            Vector.scale tileSize model.hero.position
+
+        ( xOff, yOff ) =
+            ( windowSize.width // 2, windowSize.height // 2 )
+
+        tolerance = tileSize * 4
+
+        scrollUp =
+            viewportY + y <= tolerance
+
+        scrollDown =
+            viewportY + y >= (windowSize.height * 4 // 5) - tolerance
+
+        scrollLeft =
+            viewportX + x <= tolerance
+
+        scrollRight =
+            viewportX + x >= windowSize.width - tolerance
+
+        areaSize =
+            Maps.currentAreaSize maps
+
+        newX =
+            if prevX /= x && (scrollLeft || scrollRight) then
+                xOff - x
+                    |> max (windowSize.width - areaSize.width * tileSize)
+                    |> min 0
```

Comment:
Fyi, there is also a `clamp : number -> number -> number -> number` function that may clean this up a bit. 

I also don't mind the double pipe as it is clear what you're doing here.